---
title: "Introduction to the edgynode package"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
fig_caption: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Introduction to the edgynode package}
  %\usepackage[UTF-8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
library(edgynode)
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```

# Table of Contents
1. [Why `edgynode`?](#why-edgynode)
2. [Installation](#installation)
3. [Performing an Example Workflow for edgynode analysis](#performing-an-example-workflow-for-edgynode-analysis)

## Why `edgynode`?

With the maturation of (single-cell) sequencing technologies across modalities, the inference of gene regulatory networks (`GRN`) is increasingly feasible. Yet, we lack a simple and powerful statistical package to benchmark GRN inference outcomes.

The `edgynode` package imports inferred gene regulatory networks and performs network statistics and network simulation procedures to investigate the topology and structure of the GRN at hand.

## Installation

```r
# install edgynode from GitHub
devtools::install_github("drostlab/edgynode")
```

## Performing an Example Workflow for edgynode analysis

### Load small example with internal dataset
```{r warning=FALSE, message=FALSE}
# library(edgynode)

# loading examples
load("data/adjacency_clean_test_3.rda")
load("data/adjacency_matrix_genie3_10.rda")
load("data/adjacency_matrix_test_3.rda")

# Now there are three networks to analyze
# adjacency_clean_test_3, adjacency_matrix_genie3_10 and adjacency_matrix_test_3

```

### Check data properties
```{r}
is_adjacency(adjacency_clean_test_3)

is_adjacency(adjacency_matrix_genie3_10)

is_adjacency(adjacency_matrix_test_3)
```

```{r}
# to convert those that are not adjacency formatted
adj_matrix_genie3_10 <- make_adjacency(adjacency_matrix_genie3_10)

adj_matrix_test_3 <- make_adjacency(adjacency_matrix_test_3)
```

### Benchmark small example with internal dataset

```{r warning=FALSE, message=FALSE}
# Example for weighted network

plot_adjacency_weights(adj_matrix_genie3_10, threshold = 3)
plot_adjacency_weights(adj_matrix_genie3_10, threshold = c("N1" = 2, "N2" = 1, "N3" = 3))

# Comparing node degrees in different networks

adj1 <- adj2 <- adj3 <- edgynode::make_adjacency(edgynode::adjacency_matrix_test_3)
adj2[1, 3] <- 5
adj3[1, 3] <- 0
adj1 <- edgynode::make_standard(adj1)
adj2 <- edgynode::make_standard(adj2)
adj3 <- edgynode::make_standard(adj3)

euclidean_distance_of_node_degrees <- metric_summary_distance(
 summary_fun = summary_node_degree,
 distance_fun = philentropy::euclidean,
 distance_args = list(testNA = FALSE)
 )
 
df <- compare_adj_list(
 adj_list = list("adj1" = adj1, "adj2" = adj2, "adj3" = adj3), 
 metric = euclidean_distance_of_node_degrees
 )
 
plot_compare_adj_list(df)
```

